cql3/restrictions/statement_restrictions.cc

```diff
@@ -1003,5 +1006,789 @@ ::shared_ptr<single_column_restriction> single_column_restriction::LIKE::apply_t
+    };
+    const auto zero_if_lower_one_if_upper = get_idx(bnd);
+    return boost::algorithm::any_of_equal(operators[zero_if_lower_one_if_upper], op);
+}
+
+const value_set empty_value_set = value_list{};
+const value_set unbounded_value_set = nonwrapping_range<bytes>::make_open_ended_both_sides();
+
+struct intersection_visitor {
+    const abstract_type* type;
+    value_set operator()(const value_list& a, const value_list& b) const {
+        value_list common;
+        common.reserve(std::max(a.size(), b.size()));
+        boost::set_intersection(a, b, back_inserter(common), type->as_less_comparator());
+        return std::move(common);
+    }
+
+    value_set operator()(const nonwrapping_range<bytes>& a, const value_list& b) const {
+        const auto common = b | filtered([&] (const bytes& el) { return a.contains(el, type->as_tri_comparator()); });
+        return value_list(common.begin(), common.end());
+    }
+
+    value_set operator()(const value_list& a, const nonwrapping_range<bytes>& b) const {
+        return (*this)(b, a);
+    }
+
+    value_set operator()(const nonwrapping_range<bytes>& a, const nonwrapping_range<bytes>& b) const {
+        const auto common_range = a.intersection(b, type->as_tri_comparator());
+        return common_range ? *common_range : empty_value_set;
+    }
+};
+
+value_set intersection(value_set a, value_set b, const abstract_type* type) {
+    return std::visit(intersection_visitor{type}, std::move(a), std::move(b));
+}
+
+bool is_satisfied_by(const binary_operator& opr, const column_value_eval_bag& bag) {
+    return std::visit(overloaded_functor{
+            [&] (const std::vector<column_value>& cvs) {
+                if (*opr.op == operator_type::EQ) {
+                    return equal(opr.rhs, cvs, bag);
+                } else if (*opr.op == operator_type::NEQ) {
+                    return !equal(opr.rhs, cvs, bag);
+                } else if (opr.op->is_slice()) {
+                    return limits(opr, bag);
+                } else if (*opr.op == operator_type::CONTAINS) {
+                    return contains(opr.rhs->bind_and_get(bag.options), cvs, bag);
+                } else if (*opr.op == operator_type::CONTAINS_KEY) {
+                    return contains_key(cvs, opr.rhs->bind_and_get(bag.options), bag);
+                } else if (*opr.op == operator_type::LIKE) {
+                    return like(cvs, *opr.rhs, bag);
+                } else if (*opr.op == operator_type::IN) {
+                    return is_one_of(cvs, *opr.rhs, bag);
+                } else {
+                    throw exceptions::unsupported_operation_exception("Unhandled binary_operator");
+                }
+            },
+            [] (const token& tok) -> bool {
+                // The RHS value was already used to ensure we fetch only rows in the specified
+                // token range.  It is impossible for any fetched row not to match now.
+                return true;
```

Comment:
So, after we split expressions into those representable by read_command and those used for filtering, this token expression will disappear (or it can reappear if we decide to query by index first and by token second).